doc/fbautostart.1.in
'\" t
.\"     Title: fbautostart
.\"    Author: Paul Tagliamonte <paultag@ubuntu.com>
.\" Generator: DocBook XSL Stylesheets v1.76.1 <http://docbook.sf.net/>
.\"      Date: 7th September 2011
.\"    Manual: fbautostart Manual
.\"    Source: fbautostart.txt
.\"  Language: English
.\"
.TH "FBAUTOSTART" "1" "7th September 2011" "fbautostart\&.txt" "fbautostart Manual"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
fbautostart \- Autostart all XDG applications
.SH "SYNOPSIS"
.sp
\fBfbautostart\fR [\fIoptions\fR]
.SH "DESCRIPTION"
.sp
\fBfbautostart(1)\fR is an XDG compliant startup tool\&. This will go through the directories set up by the spec, and start all applications that match its current identifier\&. You may change that identifier with the \fBFBXDG_DE\fR env var\&.
.sp
This tool has almost no library dependencies and was written to be smart about that sort of stuff\&. There is a greater then average chance that you will find a bug, since it is a from\-scratch implementation of the \&.desktop spec\&. Please report bugs\&. Please\&.
.SH "FILES"
.PP
\fB~/\&.config/autostart\fR
.RS 4
This folder contains the default user XDG dot\-desktop files that will be run ( unless it\(cqs overridden by an ENV variable )
.RE
.PP
\fB/etc/xdg\fR
.RS 4
This folder contains the default system XDG dot\-desktop files that will be run ( unless it\(cqs overridden by an ENV variable )
.RE
.SH "OPTIONS"
.sp
Sadly, there are no flags in this particular release\&.
.SH "ENVIRONMENT VARIABLES"
.PP
\fBXDG_CONFIG_HOME\fR
.RS 4
The XDG root directory for the user\(cqs startup files, which may override the global startup files\&.
.RE
.PP
\fBXDG_CONFIG_DIRS\fR
.RS 4
The root XDG directory, where global XDG files are maintained\&.
.RE
.PP
\fBFBXDG_DE\fR
.RS 4
Desktop environment that we should act upon\&. This defaults to "FLUXBOX", since this is a fluxbox project, after all\&. Please note this is case sensitive\&.
.RE
.PP
\fBFBXDG_EXEC\fR
.RS 4
This dictates if we should execute the application that we\(cqve been told to or not\&. This is particularly useful for debugging the app\&.
.sp
.if n \{\
.RS 4
.\}
.nf
Set this value to 0 to execute, and 1 to play dummy\&. I realize how silly
this is, and this behavior will likely change (with a deprecation of this
env\-var, so don\*(Aqt worry about this falling out from under your feet\&.)
.fi
.if n \{\
.RE
.\}
.RE
.PP
\fBHOME\fR
.RS 4
The location of the user\(cqs home directory, such that ~/ will be expanded to /home/user/
.RE
.SH "EXAMPLE"
.sp
.if n \{\
.RS 4
.\}
.nf
fbautostart
.fi
.if n \{\
.RE
.\}
.SH "AUTHORS"
.sp
Paul Tagliamonte <\m[blue]\fBpaultag@ubuntu\&.com\fR\m[]\&\s-2\u[1]\d\s+2> for fbautostart 2\&.718281
.SH "SEE ALSO"
.sp
fluxbox(1)
.SH "AUTHOR"
.PP
\fBPaul Tagliamonte\fR <\&paultag@ubuntu\&.com\&>
.RS 4
Author.
.RE
.SH "NOTES"
.IP " 1." 4
paultag@ubuntu.com
.RS 4
\%mailto:paultag@ubuntu.com
.RE


ChangeLog
2012-06-17  Paul Tagliamonte  <paultag@fluxbox.org>

* ERRATTUM: The fix was for gcc-4.7, not 2.7.

* src/main.cc: Properly handle paths with a ":" in them (thanks for the
report, confluence!) - Related to bug #2 on GitHub.
* maintainer/*: New directory. This contains a properly set up XDG root to
use for some basic testing. Might come in handy later, so I've included it in
the git tree.
* src/xdg_parse.cc: Check for the key before attempting to insert - std::map
will fail to insert if the key already exists, we want to overwrite.
* Version bump to 2.718281828

2012-04-03  Paul Tagliamonte  <paultag@fluxbox.org>

* src/main.cc: Including unistd.h, since we use both fork and execl
* Version bump to 2.71828182, bugfix for gcc-2.7. Thanks for the report,
Matthias Klose (doko).

2012-03-16  Paul Tagliamonte  <paultag@fluxbox.org>

* src/xdg_parse.cc(parse_folder): Adjusted this function to only parse
.desktop files. This closes bug #1 on Github. Thanks, confluence.
* src/xdg_parse.cc(ends_with): Added this function.
* Changed all the headers to show my fluxbox email rather then no email.
* Added documentation blocks to all non-straight forward routines. I'm tired
of not remembering what my old code does :)
* Added a HACKING document to help people work on the code (or myself).
* Version bump to 2.7182818.

2012-03-08  Paul Tagliamonte  <paultag@fluxbox.org>

* Re-indenteded all the files to use spaces rather then tabs.
* Added some documentation on what the FB_EXEC flag does.

2011-09-07  Paul Tagliamonte  <paultag@ubuntu.com>

* Merged with test code I've been tracking.
* Upstream version bump to version  2.718281.
* Basically, we've moved to a state-based parsing
model (without ragel, since ragel sucks ass, I don't
care what ak|ra thinks -- in the nicest way I can say
it).

2010-11-28  Paul Tagliamonte  <paultag@ubuntu.com>

* Removed the #define for the window manager setting in
include/fbautostart.h. It's now a const char, and dynamic
with the new --window-manager flag.
* Changed fbautostart to error out when you pass in a bum
flag. This caused me great pain the other day. Fixed now.
* Forgot to bump the version number. It's now version 2.71828.
Both of these changes were in src/fbautostart.cpp

2010-10-21  Paul Tagliamonte  <paultag@ubuntu.com>

* Bugfixes for random crap for XDG env issues.
This is a fix for issues that I missed when fixing
Segaja's issues.

2010-10-15  Paul Tagliamonte  <paultag@ubuntu.com>

* Bugfixes for random crap. Rewrote stuff to make
more sense.
* Fixed a few XDG spec issues. Should be betters now.
* Fixed a bug with env vars. Thanks, Segaja.

2010-10-14  Paul Tagliamonte  <paultag@ubuntu.com>

* Adding in the basic code stuff. Initial Release.


src/fbautostart.cpp
/*
 * fbautostart.cpp
 * 
 * This file is part of the Fluxbox Autostart ( fbautostart )
 * Utility.
 *
 * Copyright (C) 2011 by Paul Tagliamonte <paultag@ubuntu.com>
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

#include <iostream>
#include <vector>
#include <string>

#include "fbautostart.h"
#include "dot_desktop.cpp"

#define _APPNAME_ PACKAGE
#define _VERSION_ PACKAGE_VERSION

using namespace std;

bool noexec = false;

void version() {
	std::cout << "" << _APPNAME_ << " " << _VERSION_ << std::endl;
}

void lecture() {
	version();
	std::cout << "Copyright (C) 2010 Paul Tagliamonte" << std::endl;
	std::cout << "License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>" << std::endl;
	std::cout << "This is free software: you are free to change and redistribute it." << std::endl;
	std::cout << "There is NO WARRANTY, to the extent permitted by law." << std::endl;
}

void help() {
	std::cout << "Usage: fbautostart [OPTION] ... " << std::endl;
	std::cout << "Startup all applications according to the" << std::endl;
	std::cout << "XDG Spec. Right now, I'm hardcoded to think that" << std::endl;
	std::cout << "I am doing this on behalf of '" << _ON_BEHALF_OF << "'" << std::endl;

	std::cout << std::endl;

	std::cout << "--license             Print out license information" << std::endl;
	std::cout << "--help                Print out this message" << std::endl;
	std::cout << "--noexec              Don't exec, just do a dry run" << std::endl;
	std::cout << "--window-manager ID   Override hardcoded value ( " << _ON_BEHALF_OF <<  " )" << std::endl;

	std::cout << "" << std::endl;

	std::cout << "Copyleft (c) Paul Tagliamonte, 2010, GNU GPLv3+" << std::endl;

	std::cout << "" << std::endl;
}

int runCommand( std::string appl ) {
	/*
	 * The following is almost directly ripped off from
	 * Fluxbox's implementation. ( at ak|ra's behest )
	 */

	pid_t pid = fork();
	if (pid)
		return pid;

	// get shell path from the environment
	//
	// this process exits immediately, so we don't have to
	//  worry about memleaks
	std::string shell = getenv("SHELL");

	if ( shell != "" )
		shell = "/bin/sh";

	if ( ! noexec ) { // we'll do it live
		execl(
			shell.c_str(),
			shell.c_str(),
			"-c",
			appl.c_str(),
			static_cast<void*>(NULL)
		);
		exit ( EXIT_SUCCESS );
		return pid; // compiler happy -> we are happy ;)
	} else { // dummy mode ( test )
		std::cout << "Would have run: " << appl << std::endl;
		exit(0);
		return 0;
	}
}

void processArgs( int argc, char ** args ) {
	for ( int i = 1; i < argc; ++i ) {
		if ( strcmp(args[i], "--license") == 0 ) {
			lecture();
			exit(0);
		} else if ( strcmp(args[i], "--version") == 0 ) {
			version();
			exit(0);
		} else if ( strcmp(args[i], "--help") == 0 ) {
			help();
			exit(0);
		} else if ( strcmp(args[i], "--noexec") == 0 ) {
			noexec = true;
		} else if ( strcmp(args[i], "--window-manager") == 0 ) {
			_ON_BEHALF_OF = args[i+1];
			i = i+1;
		} else {
			help();

			std::cout << "Holy god. I have no idea what:" << std::endl;
			std::cout << "  " << args[i] << std::endl;
			std::cout << "means. Try again :/" << std::endl;
			std::cout << std::endl;

			exit(1);
		}
	}
}

int main ( int argc, char ** argv ) {
	processArgs( argc, argv );

	if ( noexec ) {
		std::cout << "Warning: In noexec mode." << std::endl;
	}

	std::cout << "Launching on behalf of " << _ON_BEHALF_OF << std::endl;

	std::vector<dot_desktop>   files;
	std::vector<std::string>   dirs;

	if ( getConfDirs( dirs ) ) { // if no directories barf in our face
		if ( getDesktopFiles( dirs, files ) ) { // and we load everything with glee
			for ( unsigned int i = 0; i < files.size(); ++i ) { // run through all the files
				dot_desktop d = files.at(i);
				bool happy = true;
				std::string only = d.getAttr("OnlyShowIn"); // Only one per file ( per xdg )
				std::string noti = d.getAttr("NotShowIn");  // We'll ignore that until we care
				                                             // XXX: This is most likely a bug.
				if ( only != "" ) { // even if an attr does not exist
				                    // the object will return it as "".
					int index = -1;
					index = only.find(_ON_BEHALF_OF); // if we have our WM in the OnlyLaunch
					if ( index < 0 ) { // we're disabled ( not found )
						happy = false;
						debug("");
						debug("Not running the following app ( Excluded by a OnlyShowIn )");
						debug(d.getAttr("Name"));
					}
				}
				if ( noti != "" ) { // (NotShowIn (don't show))
					int index = -1;
					index = noti.find(_ON_BEHALF_OF); // if we have found our WM
					if ( index >= 0 ) { // We're in Launch, stop from launching it.
						happy = false;
						debug("");
						debug("Forced into not running the following app ( Included by not being in NotShowIn )");
						debug(d.getAttr("Name"));
					}
				}
				if ( d.getAttr("Hidden") == "" && happy ) { // If we sould exec
					std::string appl = d.getAttr("Exec"); // get the line to run
					if ( appl != "" ) { // if it's defined and ready to go
						debug( "Processing File: ");
						debug(d.getFile());
						runCommand( appl ); // kickoff (regardless of noexec)
					}
				} // otherwise, we're out of here.
			}
			return 0;
		}
		return 0xDEADBEEF;
	}
	return 0xCAFEBABE;
}

